tensornetwork/tests/testing_utils.py

```diff
@@ -0,0 +1,108 @@
+import numpy as np
+import jax
+import jax.numpy as jnp
+from jax import config
+import tensorflow as tf
+import torch
+import pytest
+import tensornetwork
+from tensornetwork import backends, backend_contextmanager
+config.update("jax_enable_x64", True)
+
+np_real = [np.float32, np.float64]
+np_complex = [np.complex64, np.complex128]
+np_float_dtypes = np_real + np_complex
+np_int = [np.int8, np.int16, np.int32, np.int64]
+np_uint = [np.uint8, np.uint16, np.uint32, np.uint64]
+np_not_bool = np_float_dtypes + np_int + np_uint + [None,]
+np_not_half = [np.float32, np.float64] + np_complex
+np_all_dtypes = np_not_bool + [np.bool,]
+
+torch_supported_dtypes = np_real + np_int + [np.uint8, np.bool, None]
+#torch_supported_dtypes = [np.float32, np.float64]
+
+def safe_randn(shape, backend, dtype):
+  """
+  Creates a random tensor , catching errors that occur when the
+  dtype is not supported by the backend. Returns the Tensor and the backend
+  array, which are both None if the dtype and backend did not match.
+  """
+  np.random.seed(seed=10)
+  init = np.random.randn(*shape)
+  if dtype == np.bool:
+    init = np.round(init)
+  init = init.astype(dtype)
+
+  if dtype in np_complex:
+    init_i = np.random.randn(*shape)
+    init = init + 1.0j * init_i.astype(dtype)
+
+  if backend == "pytorch" and dtype not in torch_supported_dtypes:
+    pytest.skip("dtype unsupported by PyTorch")
+    A = None
+    init = None
+  else:
+    A = tensornetwork.Tensor(init, backend=backend)
+  return (A, init)
+
+
+def safe_zeros(shape, backend, dtype):
+  """
+  Creates a tensor of zeros, catching errors that occur when the
+  dtype is
+  not supported by the backend. Returns both the Tensor and the backend array,
+  which are both None if the dtype and backend did not match.
+  """
+  init = np.zeros(shape, dtype=dtype)
+  if backend == "pytorch" and dtype not in torch_supported_dtypes:
+    pytest.skip("dtype unsupported by PyTorch")
+    A = None
+    init = None
+  else:
+    A = tensornetwork.Tensor(init, backend=backend)
+  return (A, init)
+
+
+def np_dtype_to_backend(backend, dtype):
+  """
+  Converts a given np dtype to the equivalent in the given backend. Skips
+  the present test if the dtype is not supported in the backend.
+  """
+  if backend is None:
+    backend = backend_contextmanager.get_default_backend()
```

Comment:
This line is never reached. You can reach this by having a test use the `with tn.DefaultBackend(backend=...):` context manager instead of directly setting the backend yourself.